lib/charms/opensearch/v0/opensearch_performance_profile.py

```diff
@@ -0,0 +1,240 @@
+                                 switched the profile
+
+The charm will then apply the profile and restart the OpenSearch service if needed.
+"""
+import logging
+import os
+
+import ops
+from charms.opensearch.v0.constants_charm import (
+    PERFORMANCE_PROFILE,
+    PeerClusterRelationName,
+    PeerRelationName,
+)
+from charms.opensearch.v0.models import (
+    DeploymentType,
+    OpenSearchPerfProfile,
+    PerformanceType,
+)
+from charms.opensearch.v0.opensearch_exceptions import OpenSearchHttpError
+from charms.opensearch.v0.opensearch_internal_data import Scope
+from ops.framework import EventBase, EventSource
+
+# The unique Charmhub library identifier, never change it
+LIBID = "8b7aa39016e748ea908787df1d7fb089"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+logger = logging.getLogger(__name__)
+
+
+class _ApplyProfileTemplatesOpenSearch(EventBase):
+    """Attempt to acquire lock & restart OpenSearch.
+
+    The main reason to have a separate event, is to be able to wait for the cluster.
+    It defers otherwise and only defers the execution of this particular task.
+    """
+
+
+class OpenSearchPerformance(ops.Object):
+    """Base class for OpenSearch charms."""
+
+    _apply_profile_templates_event = EventSource(_ApplyProfileTemplatesOpenSearch)
+
+    def __init__(self, charm: ops.charm.CharmBase = None):
+        super().__init__(charm, None)
+        self.charm = charm
+        self.peers_data = self.charm.peers_data
+
+        self.framework.observe(
+            charm.on.install,
+            self._on_install,
+        )
+        self.framework.observe(
+            charm.on[PeerClusterRelationName].relation_changed,
+            self._on_peer_cluster_relation_changed,
+        )
+        self.framework.observe(
+            charm.on[PeerRelationName].relation_changed,
+            self._on_peer_cluster_relation_changed,
+        )
+        self.framework.observe(
+            self._apply_profile_templates_event, self._on_apply_profile_templates
+        )
+        self._apply_profile_templates_has_been_called = False
+
+    def _on_install(self, event: EventBase):
+        """Handle install event.
+
+        Execute it on install, as defined in the docs, it is the first hook to run:
+           https://juju.is/docs/sdk/config-changed-event
+
+        install -> config-changed -> start
+
+        Store the current perf. profile we are applying. This must happen solely
+        on the beginning of the charm lifecycle. Therefore, we check if the unit is down.
+        """
+        if not os.path.exists(
+            os.path.join(
+                self.charm.opensearch.paths.conf,
+                self.charm.opensearch_config.JVM_OPTIONS,
+            )
+        ):
+            event.defer()
+            return
```

Comment:
no need for this. This comes for free when installing, by the time you're in config_changed you are good.